src/Storages/StorageMerge.cpp

```diff
@@ -666,6 +721,17 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
                 processed_stage,
                 max_block_size,
                 UInt32(streams_num));
+
+            if (!plan.isInitialized())
+                return {};
+
+            if (row_policy_data.needCare())
+            {
+                if (auto * source_step_with_filter = dynamic_cast<SourceStepWithFilter*>((plan.getRootNode()->step.get())))
```

Comment:
What if that `step` is not derived from `SourceStepWithFilter`?

Reply:
Please add a comment here that it's an optimization if the storage can use filtering while reading (e.g. MergeTree).

Reply:
The conditions can be combined to be shorter:

```
if (plan.isInitialized() and row_policy_data)
{
    if (auto * source_step_with_filter = dynamic_cast<SourceStepWithFilter*>((plan.getRootNode()->step.get())))
        row_policy_data->addStorageFilter(source_step_with_filter);
}
```

Reply:
It is not obvious that "The conditions can be combined to be shorter".
Could you explain why you are sure that going further with non initialized _plan_ is safe and makes sense?